gremlin-go/driver/graphTraversal.go

```diff
@@ -670,3 +675,99 @@ func (g *GraphTraversal) Write(args ...interface{}) *GraphTraversal {
 	g.bytecode.addStep("write", args...)
 	return g
 }
+
+type transaction struct {
+	g                      *GraphTraversalSource
+	sessionBasedConnection *DriverRemoteConnection
+	remoteConnection       *DriverRemoteConnection
+	isOpen                 bool
+	mutex                  sync.Mutex
+}
+
+func (t *transaction) Begin() (*GraphTraversalSource, error) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if err := t.verifyTransactionState(false, "Transaction already started on this object"); err != nil {
+		return nil, err
+	}
+
+	sc, err := t.remoteConnection.CreateSession()
+	if err != nil {
+		return nil, err
+	}
+	t.sessionBasedConnection = sc
+	t.isOpen = true
+
+	gts := &GraphTraversalSource{
+		graph:               t.g.graph,
+		traversalStrategies: t.g.traversalStrategies,
+		bytecode:            t.g.bytecode,
+		remoteConnection:    t.sessionBasedConnection}
+	return gts, nil
+}
+
+func (t *transaction) Rollback() error {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if err := t.verifyTransactionState(true, "Cannot rollback a transaction that is not started."); err != nil {
+		return err
+	}
+
+	return t.closeSession(t.sessionBasedConnection.rollback())
+}
+
+func (t *transaction) Commit() error {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if err := t.verifyTransactionState(true, "Cannot commit a transaction that is not started."); err != nil {
+		return err
+	}
+
+	return t.closeSession(t.sessionBasedConnection.commit())
+}
+
+func (t *transaction) Close() error {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	err := t.verifyTransactionState(true, "Cannot close a transaction that has previously been closed.")
+	if err != nil {
+		return err
+	}
+
+	return t.closeSession(nil, nil)
+}
+
+func (t *transaction) IsOpen() bool {
+	if t.sessionBasedConnection != nil && t.sessionBasedConnection.closed {
+		t.isOpen = false
+	}
+	return t.isOpen
+}
+
+func (t *transaction) verifyTransactionState(state bool, error string) error {
```

Comment:
I'd imagine this will need to be changed (as well other error formatting) when merging with Rithin's PR, should merging this come after merging Rithin's PR?

Reply:
done in https://github.com/lyndonbauto/tinkerpop/pull/68/commits/402e323b1867ab4460f0bcd5f0a2515f6638366a